app/cli.py

```diff
@@ -0,0 +1,235 @@
+from ocp_utilities.utils import run_command
+
+
+def get_install_config_j2_template(cluster_dict):
+    template_file = "install-config-template.j2"
+    env = Environment(
+        loader=FileSystemLoader("app/manifests/"),
+        trim_blocks=True,
+        lstrip_blocks=True,
+        undefined=DebugUndefined,
+    )
+
+    template = env.get_template(name=template_file)
+    rendered = template.render(cluster_dict)
+    undefined_variables = meta.find_undeclared_variables(env.parse(rendered))
+    if undefined_variables:
+        click.echo(f"The following variables are undefined: {undefined_variables}")
+        raise click.Abort()
+
+    return yaml.safe_load(rendered)
+
+
+def generate_cluster_dir_path(clusters, base_directory):
+    for _cluster in clusters:
+        _cluster["install-dir"] = os.path.join(
+            base_directory, _cluster["platform"], _cluster["name"]
+        )
+    return clusters
+
+
+def verify_processes_passed(processes, action):
+    failed_processes = {}
+
+    for _proc in processes:
+        _proc.join()
+        if _proc.exitcode != 0:
+            failed_processes[_proc.name] = _proc.exitcode
+
+    if failed_processes:
+        click.echo(f"Some jobs failed to {action}: {failed_processes}\n")
+        raise click.Abort()
+
+
+def download_openshift_install_binary(clusters, pull_secret_file):
+    versions = set()
+    openshift_install_str = "openshift-install"
+
+    for cluster in clusters:
+        versions.add(cluster["version"])
+
+    for version in versions:
+        binary_dir = os.path.join("/tmp", version)
+        clusters = [_cluster for _cluster in clusters if _cluster["version"] == version]
+        for cluster in clusters:
+            cluster["openshift-install-binary"] = os.path.join(
+                binary_dir, openshift_install_str
+            )
+
+        run_command(
+            command=shlex.split(
+                "oc adm release extract "
+                f"quay.io/openshift-release-dev/ocp-release:{version}-x86_64 "
+                f"--command={openshift_install_str} --to={binary_dir} --registry-config={pull_secret_file}"
+            ),
+            check=False,
+        )
+
+    return clusters
+
+
+def create_or_destroy_cluster(cluster_data, action):
+    directory = cluster_data["install-dir"]
+    binary_path = cluster_data["openshift-install-binary"]
+    return run_command(
+        command=shlex.split(f"{binary_path} {action} cluster --dir {directory}"),
+        capture_output=False,
+    )[0]
+
+
+def install_openshift(cluster_data):
+    create_or_destroy_cluster(cluster_data=cluster_data, action="create")
+
+
+def uninstall_openshift(cluster_data):
+    create_or_destroy_cluster(cluster_data=cluster_data, action="destroy")
+
+
+def create_install_config_file(clusters, pull_secret_file):
+    ssh_key = get_local_ssh_key()
+    pull_secret = json.dumps(get_pull_secret_data(pull_secret_file=pull_secret_file))
```

Comment:
As I understand the cli is getting a pull secret file indication and not reffering it to a specific cluster - is it "forcing" all cluster to be installed with the specified pull secret?

Reply:
the pull secret is to get the images from redhat registry